corfudb-tools/src/main/java/org/corfudb/compactor/CorfuStoreCompactorMain.java

```diff
@@ -0,0 +1,336 @@
+                    runtimeTrustStore, runtimeTrustStorePasswordFile);
+
+        } else {
+            corfuRuntimeHelper = new CorfuRuntimeHelper(hostname, port, maxWriteSize, bulkReadSize);
+            cpRuntimeHelper = new CorfuRuntimeHelper(hostname, port, maxWriteSize, bulkReadSize);
+        }
+
+        corfuRuntime = corfuRuntimeHelper.getRuntime();
+        cpRuntime = cpRuntimeHelper.getRuntime();
+        corfuStore = new CorfuStore(corfuRuntime);
+        distributedCompactor = new DistributedCompactor(corfuRuntime, cpRuntime, persistedCacheRoot);
+
+        this.openCompactionTables();
+    }
+
+    private void upgrade() {
+        retryCheckpointing = 10;
+        if (upgradeDescriptorTable) {
+            syncProtobufDescriptorTable();
+        }
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            txn.putRecord(checkpointTable, DistributedCompactor.UPGRADE_KEY, TokenMsg.getDefaultInstance(),
+                    null);
+            txn.commit();
+        } catch (Exception e) {
+            log.warn("Unable to write UpgradeKey to checkpoint table, ", e);
+        }
+    }
+
+    /**
+     * Entry point to invoke Client checkpointing by the CorfuServer
+     *
+     * @param args command line argument strings
+     */
+    @SuppressWarnings("UncommentedMain")
+    public static void main(String[] args) throws Exception {
+        getCompactorArgs(args);
+
+        if (maxWriteSize == -1) {
+            if (persistedCacheRoot == null) {
+                // in-memory compaction
+                maxWriteSize = DEFAULT_CP_MAX_WRITE_SIZE;
+            } else {
+                // disk-backed non-config compaction
+                maxWriteSize = NON_CONFIG_DEFAULT_CP_MAX_WRITE_SIZE;
+            }
+        }
+
+        if (port == 9040) {
+            Thread.currentThread().setName("CS-NonConfig-chkpter");
+        } else {
+            Thread.currentThread().setName("CS-Config-chkpter");
+        }
+
+        CorfuStoreCompactorMain corfuCompactorMain = new CorfuStoreCompactorMain();
+
+        if (isUpgrade) {
+            corfuCompactorMain.upgrade();
+        }
+        corfuCompactorMain.checkpoint();
+    }
+
+    private void checkpoint() {
+        try {
+            for (int i = 0; i< retryCheckpointing; i++) {
+                if (distributedCompactor.startCheckpointing() > 0) {
+                    break;
+                }
+                TimeUnit.MILLISECONDS.sleep(1000);
+            }
+        } catch (Throwable throwable) {
+            log.error("CorfuStoreCompactorMain crashed with error:", CORFU_LOG_CHECKPOINT_ERROR, throwable);
+        }
+    }
+
+    private void openCompactionTables() {
+        try {
+            checkpointTable = corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
+                    DistributedCompactor.CHECKPOINT,
+                    StringKey.class,
+                    TokenMsg.class,
+                    null,
```

Comment:
why null?

Reply:
null in metadata section means there is no metadata in addition to whatever the payload has.
These tables are already checkpoint metadata tables and we need to keep the state here as small as possible, so we avoid defining this section all together.